### PR TITLE
feat (Telemetry) Use Summary to record percentile metrics of time

### PR DESCRIPTION
**Pull request by gordonbear**
Tested locally by using the updated offchain-sdk to record latency metrics of BTS-indexer. So far so good. See the screenshots.

Screenshot 2024-06-18 at 1 51 02 PM

### CODE REVIEW

File: telemetry/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+* `StatsdAddr`: The address of the Datadog StatsD client. This is needed if the metrics should be
+emitted from Datadog.
+
+* `Namespace`: This will appear as the `Namespace` tag in Datadog.
+
+### Datadog Methods
+
+[metrics.go](./datadog/metrics.go) implements the Datadog version of the supported metrics methods
+defined in [type.go](./type.go). All implementations are simple wrappers around the native methods
+in Datadog.
```
Comment (calbera) [inline telemetry/README.md:30]:
```suggestion
defined in [types.go](./types.go). All implementations are simple wrappers around the native methods
provided by the Datadog `statsd` client.
```

File: telemetry/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+## Prometheus
+
+### Configuration
+
+The first step is to add a section in your config file. The source code defining these configs can
+be found in  [config.go](./prometheus/config.go).
```
Comment (calbera) [inline telemetry/README.md:37]:
```suggestion
be found in [config.go](./prometheus/config.go).
```

File: telemetry/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+#### Prometheus Configs
+
+* `Enabled`: Set to true to enable metrics emission to Prometheus.
+
+* `Namespace` and `Subsystem`: These fields will be added as prefixes to the metrics name.
+For example, if `Namespace` is `app` and `Subsystem` is `api`, then the full metrics name of
+`request_success` will be `app_api_request_success`.
+
+* `HistogramBucketCount`: The number of buckets used for Histogram typed metrics. Default is 10.
```
Comment (calbera) [inline telemetry/README.md:47]: add a `Note:` here explaining what bucket size is recommended and what is the recommended maximum

File: telemetry/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+* `Gauge`: This method wraps the `Gauge` metrics of Prometheus.
+
+* `Decr` and `Incr`: Implemented using the `Gauge` metrics of Prometheus.
+
+* `Count`: This method wraps the `Count` metrics of Prometheus. Note that after deployment or instance
+restart, `Count` will reset to 0. This is by design in Prometheus.
+
+* `IncMonotonic` and `Error`: Implemented using the `Count` metrics of Prometheus.
+
+* `Histogram`: This method wraps the `Histogram` metrics of Prometheus with linear buckets.
```
Comment (calbera) [inline telemetry/README.md:67]: would add a todo/nit comment here saying we want to support different types of buckets in the future.

File: telemetry/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# telemetry
+
+The metrics utility for offchain-sdk.
+
+[type.go](./type.go) defines the interface for the supported metrics methods.
```
[the remark this answers is not in the document]
Comment (gordonbear) [reply telemetry/README.md:5]: Good point. I will also rename the file

File: telemetry/README.md (hunk 3 above)
Comment (gordonbear) [reply telemetry/README.md:47]: Mentioning the "recommended" bucket count here.

While the `maximum` supported range is related more to the method logic, thus mention it in the `Histogram` method

**Review by calbera — APPROVED** [review]

**Pull request merged**